kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/HighOffsetMetricsExposer.scala

```diff
@@ -0,0 +1,121 @@
+import zio.kafka.consumer.Consumer
+import zio.{Duration, Fiber, Schedule, Task, UIO, URIO, ZIO, ZLayer}
+import zio.stream.{ZSink, ZStream}
+import zio.metrics.*
+import zio.metrics.Metric.Counter
+
+import scala.collection.immutable.Map
+
+object Helper {
+  def newTopicMetric(topic: String): Metric.Counter[Long] = Metric.counter("kafkakewl_vnext_topic_high_offset").tagged("topic", topic)
+  def newPartitionMetric(topic: String, partition: Int): Metric.Counter[Long] =
+    Metric.counter("kafkakewl_vnext_topic_partition_high_offset").tagged("topic", topic).tagged("partition", partition.toString)
+}
+
+case class MetricMap(metricMap: Map[String, (Map[Int, (Metric.Counter[Long], Long)], Metric.Counter[Long])]) {
+
+  def updateWithTopicParitionChanges(
+      changes: KafkaTopicPartitionInfoChanges,
+      config: MetricExposerConfig
+  ): (MetricMap, List[(Metric.Counter[Long], Long)]) = {
+    val addedOrUpdated = changes.addedOrUpdated
+    val removed = changes.removed
+    var newMetricMap = metricMap
+    var metricUpdates: List[(Metric.Counter[Long], Long)] = List.empty
+    // Add new metrics and update stale ones
+    for ((topicPartition, partitionInfo) <- addedOrUpdated) {
+
+      newMetricMap = newMetricMap.updatedWith(topicPartition.topic) {
+        case Some((topicMap: Map[Int, (Metric.Counter[Long], Long)], aggregateCounter: Metric.Counter[Long])) =>
+          val newTopicMap = topicMap.updatedWith(topicPartition.partition) {
+            case Some((counter, lastValue)) =>
+              val increase = partitionInfo.endOffset - lastValue
+              metricUpdates = metricUpdates :+ (aggregateCounter, increase)
+              if (config.perPartitionMetrics) {
+                metricUpdates = metricUpdates :+ (counter, increase)
+              }
+              Some((counter, partitionInfo.endOffset))
+            case None =>
+              val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
+              metricUpdates = metricUpdates :+ (aggregateCounter, partitionInfo.endOffset)
+              if (config.perPartitionMetrics) {
+                metricUpdates = metricUpdates :+ (counter, partitionInfo.endOffset)
+              }
+              Some((counter, partitionInfo.endOffset))
+          }
+          Some((newTopicMap, aggregateCounter))
+        case None =>
+          val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
+          val newTopicMap = Map.newBuilder
+            .addOne(
+              topicPartition.partition,
+              (counter, partitionInfo.endOffset)
+            )
+            .result()
+          val aggregateCounter = Helper.newTopicMetric(topicPartition.topic)
+          metricUpdates = metricUpdates :+ (aggregateCounter, partitionInfo.endOffset)
+          if (config.perPartitionMetrics) {
+            metricUpdates = metricUpdates :+ (counter, partitionInfo.endOffset)
```

Comment:
The per-partition counter is created and stored even if `perPartitionMetrics` isn't true - does this mean it'll exposed with zero? Shouldn't we not create it at all if `perPartitionMetrics` isn't true?

Of course this would mean that the `Some` case above needs some changes because we won't have the per-partition counter.